extracted tag parsing logic, changes for NET6/NET7

AoshiW opened the pull request:
Changes:
- (performance) changes for NET6/NET7/netstandard2.1
- deleted all `Color Color` properties (or replaced by `string HexColor` property) and created interface for easier parsing of HexColor on user-side
- replace all `string TmiSendTs` properties with `DateTimeOffset TmiSend`
- extracted tag parsing logic into a separate class

Syzuna reviewed (commented):
I don't quite get why you changed it first just to revert it for a nicer gitdiff?

AoshiW commented:
I changed it to `tag.Value` automatically without thinking but then I realized that it was an unnecessary (and maybe even unwanted) change so I changed it back to 
(so please use `Squash and merge`  so no one sees a useless commit)

neon-sunset commented:
@AoshiW Btw, if you want to clean up git history, you can "Undo" commits in VS Code + GitLens or just reset on a specific branch, then commit again in the desired form, and then `git push -f` - this will override commit history, making it seem as if the previous commits never happened.

swiftyspiffy commented on TwitchLib.Client.Models/Builders/GiftedSubscriptionBuilder.cs:
```diff
@@ -8,7 +7,7 @@ public sealed class GiftedSubscriptionBuilder : IBuilder<GiftedSubscription>, IF
         private readonly List<KeyValuePair<string, string>> _badges = new List<KeyValuePair<string, string>>();
         private readonly List<KeyValuePair<string, string>> _badgeInfo = new List<KeyValuePair<string, string>>();
-        private string _color;
+        private Color _color;
```
so we actually had this originally at some point but it required declaring a dependency on System.Drawing, which iirc was breaking things in unity. Not sure if @Syzuna has additional context on this. If y'all think we're good here on this, its cool with me

neon-sunset replied:
IIRC Unity isn't the only thing that has issues with System.Drawing - certain other targets (and Native AOT) had certain cases where System.Drawing required replacing it with SkiaSharp.

Seconding the suggestion that it's best to try to avoid referencing it.

AoshiW replied:
if the dependency on System.Drawing is such a problem we should solve it somehow
maybe we could address this in the new Issue instead of the PR


There may be solutions:

- Ignore it and continue to use `Color` from System.Drawing
- Replace everything with the string HeColor`
- Replace everything with a custom `struct Color` string
- Replace everything with `Color` from SkiaSharp (or other alternative)

swiftyspiffy replied:
Imo I think the best route is probably just a string for now. Let the implementations using the library determine how to handle the color (presumably most will convert to System.Drawing.Color).

AoshiW replied:
what is your opinion on it @Syzuna?

AoshiW replied:
if we use a string, we could create an interface
```en
interface IHexColorProperty // this needs a better name
{
     string HexColor { get; }
}
```
so people could then easily create an extension method for parsing color on "models"

swiftyspiffy replied:
Ya, that seems reasonable I think. I'll let you decide the correct implementation, I just wanted to steer clear of System.Drawing.

swiftyspiffy commented on TwitchLib.Client.Models/ConnectionCredentials.cs:
```diff
@@ -1,11 +1,17 @@
+    public partial class ConnectionCredentials
     {
+#if NET7_0_OR_GREATER
+        [GeneratedRegex("^([a-zA-Z0-9][a-zA-Z0-9_]{3,25})$")]
```
nit: think this should be {4,25}

AoshiW replied:
fixed

swiftyspiffy reviewed (approved):
lgtm, just seeking clarity on the Color possible issue

swiftyspiffy commented:
lgtm
